### PR TITLE
Improve testing of mismatched field numbers.

### LINKED ISSUE #13805: TestLucene90DocValuesFormat fails with ArrayIndexOutOfBoundsException
```
ERROR: The following test(s) have failed:
  - org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields (:lucene:core)
    Test output: /opt/buildkite-agent/builds/bk-agent-prod-gcp-1726674638633683811/elastic/apache-lucene-nightly/lucene/core/build/test-results/test/outputs/OUTPUT-org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.txt
    Reproduce with: gradlew :lucene:core:test --tests "org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields" -Ptests.jvms=12 -Ptests.jvmargs= -Ptests.seed=175AB2293A24B66E -Ptests.nightly=true -Ptests.gui=true -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=128 -Ptests.forceintegervectors=true
  - org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.testSparseDocValuesVsStoredFields (:lucene:backward-codecs)
    Test output: /opt/buildkite-agent/builds/bk-agent-prod-gcp-1726674638633683811/elastic/apache-lucene-nightly/lucene/backward-codecs/build/test-results/test/outputs/OUTPUT-org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.txt
    Reproduce with: gradlew :lucene:backward-codecs:test --tests "org.apache.lucene.backward_codecs.lucene80.TestBestSpeedLucene80DocValuesFormat.testSparseDocValuesVsStoredFields" -Ptests.jvms=12 -Ptests.jvmargs= -Ptests.seed=175AB2293A24B66E -Ptests.nightly=true -Ptests.gui=true -Ptests.file.encoding=UTF-8 -Ptests.vectorsize=128 -Ptests.forceintegervectors=true
```

```
 >     java.lang.ArrayIndexOutOfBoundsException: Index 3 out of bounds for length 3
   >         at __randomizedtesting.SeedInfo.seed([175AB2293A24B66E:43816E3AC0A89021]:0)
   >         at org.apache.lucene.util.packed.Packed64.get(Packed64.java:80)
   >         at org.apache.lucene.index.OrdinalMap$1.get(OrdinalMap.java:379)
   >         at org.apache.lucene.codecs.DocValuesConsumer$7$1.nextOrd(DocValuesConsumer.java:946)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer$4$1.nextDoc(Lucene90DocValuesConsumer.java:808)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.writeValues(Lucene90DocValuesConsumer.java:201)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.doAddSortedNumericField(Lucene90DocValuesConsumer.java:705)
   >         at org.apache.lucene.codecs.lucene90.Lucene90DocValuesConsumer.addSortedSetField(Lucene90DocValuesConsumer.java:770)
   >         at org.apache.lucene.codecs.DocValuesConsumer.mergeSortedSetField(DocValuesConsumer.java:853)
   >         at org.apache.lucene.codecs.DocValuesConsumer.merge(DocValuesConsumer.java:148)
   >         at org.apache.lucene.codecs.perfield.PerFieldDocValuesFormat$FieldsWriter.merge(PerFieldDocValuesFormat.java:152)
   >         at org.apache.lucene.index.SegmentMerger.mergeDocValues(SegmentMerger.java:188)
   >         at org.apache.lucene.index.SegmentMerger.mergeWithLogging(SegmentMerger.java:314)
   >         at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:149)
   >         at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:5292)
   >         at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:4758)
   >         at org.apache.lucene.index.IndexWriter$IndexWriterMergeSource.merge(IndexWriter.java:6581)
   >         at org.apache.lucene.index.SerialMergeScheduler.merge(SerialMergeScheduler.java:38)
   >         at org.apache.lucene.index.IndexWriter.executeMerge(IndexWriter.java:2327)sFormat
   >         at org.apache.lucene.index.IndexWriter.maybeMerge(IndexWriter.java:2322)
   >         at org.apache.lucene.index.IndexWriter.processEvents(IndexWriter.java:6033)
   >         at org.apache.lucene.index.IndexWriter.maybeProcessEvents(IndexWriter.java:6023)
   >         at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1562)
   >         at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1847)
   >         at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1487)
   >         at org.apache.lucene.tests.index.RandomIndexWriter.addDocument(RandomIndexWriter.java:224)
   >         at org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.doTestSparseDocValuesVsStoredFields(TestLucene90DocValuesFormat.java:215)
   >         at org.apache.lucene.codecs.lucene90.TestLucene90DocValuesFormat.testSparseDocValuesVsStoredFields(TestLucene90DocValuesFormat.java:169)
   >         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:580)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
   >         at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
   >         at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
   >         at org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
   >         at java.base/java.lang.Thread.run(Thread.java:1583)
```

### PR DESCRIPTION
**Pull request by jpountz**
This improves testing of mismatched field numbers by
 - improving `AssertingDocValuesProducer` to detect mismatched field numbers,
 - introducing a `MismatchedCodecReader` to actually test mismatched field numbers on `DocValuesProducer` (a `MismatchedLeafReader` wrapping a `SlowCodecReaderWrapper` doesn't work since `SlowCodecReaderWrapper` implicitly resolves the correct `FieldInfo` object),
 - introducing an explicit test for mismatched field numbers in `BaseDocValuesFormatTestCase`.

These new tests uncovered a bug when merging sorted doc values, which would call the underlying doc values producer with the merged field info.

Closes #13805

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/DocValuesConsumer.java [hunk 1]
```diff
@@ -613,7 +613,7 @@ public void mergeSortedField(FieldInfo fieldInfo, final MergeState mergeState)
       if (docValuesProducer != null) {
         FieldInfo readerFieldInfo = mergeState.fieldInfos[i].fieldInfo(fieldInfo.name);
         if (readerFieldInfo != null && readerFieldInfo.getDocValuesType() == DocValuesType.SORTED) {
-          values = docValuesProducer.getSorted(fieldInfo);
+          values = docValuesProducer.getSorted(readerFieldInfo);
```
Comment (ChrisHegarty) [inline lucene/core/src/java/org/apache/lucene/codecs/DocValuesConsumer.java:616]: 👍

File: lucene/test-framework/src/java/org/apache/lucene/tests/codecs/asserting/AssertingDocValuesFormat.java [hunk 2]
```diff
@@ -229,6 +234,7 @@ static class AssertingDocValuesProducer extends DocValuesProducer {
 
     @Override
     public NumericDocValues getNumeric(FieldInfo field) throws IOException {
+      assert fieldInfos.fieldInfo(field.name).number == field.number;
```
Comment (ChrisHegarty) [inline lucene/test-framework/src/java/org/apache/lucene/tests/codecs/asserting/AssertingDocValuesFormat.java:237]: ++

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by iverase** [comment]
The failure seems legit. It happens because of this method in MismatchedCodecReader:

```
  @Override
  public FieldInfos getFieldInfos() {
    return shuffled;
  }
```

if you replace this with return in.getFieldInfos() then it is successful. I think the method FiedInfos#getMergedFieldInfos get confused because of soft deletes and parent fields?

**Comment by jpountz** [comment]
We haven't had failures in thes test case for a very long time, so I suspect it's a genuine failure indeed, which got triggered by the additional testing of mismatched field numbers?

**Comment by rmuir** [comment]
Seems the test/functionality uses points and docvalues. Original failing test that inspired this PR  `testSparseDocValuesVsStoredFields`. Maybe we need a similar one to cover DocValuesVsPoints?

I guess my concern here is that fixes/test heres are focused on docvalues, but fieldinfo numbers are used by all index structures and there might be bugs elsewhere too... just do they have the tests to find issues?

Thanks for digging into the field numbers issues.

**Pull request merged**